Add PR template specific to the schema soft freeze (`berkeley-schema-fy24`)

mslarae13 opened the pull request:
trial and error, so many commits.

See branch to preview the PR template: https://github.com/microbiomedata/berkeley-schema-fy24/blob/1995-add-pr-template_berk_schema/.github/PULL_REQUEST_TEMPLATE/pull_request_template_berk.md

Add comments to the .md file. 
the .yml file won't work for PRs

turbomam commented on .github/ISSUE_TEMPLATE/pull_request_template_berk_schema.yml:
```diff
@@ -0,0 +1,103 @@
+      value: |
+        # Guidelines
+        
+        ## _Light_ Schema Freeze
```
I think we have many references to this as a soft freeze in other documents

turbomam commented on .github/ISSUE_TEMPLATE/pull_request_template_berk_schema.yml:
```diff
@@ -0,0 +1,103 @@
+        ## Reviewers
+        To ensure no changes are made unexpectedly, ALL berkeley-schema refacror roll out task coordinators should be listed as reviewrs for ALL PRs. 
+        See [squad proposal](https://docs.google.com/document/d/1XXN1YuaBuSkxPXeiLKm5YxYzXTamBPQrzzeLhlh7PWs/edit#heading=h.u52g8v319adh) for list of names.
+        PRs should NOT be merged until they have been approved by all task coordinators.
```
What commitment are the task leaders making to punctually review PRs?

turbomam replied:
Is this an issue template a PR template or both?

mslarae13 replied:
It's a PR template but far as I could tell, they all go into the ISSUE_TEMPLATE folder. but I'm gonna have to deploy it and check!

mslarae13 replied:
made PR template directory

turbomam reviewed (commented):
this will be super helpful. i may think of  a few more revision suggestions in the next 24 hours.

turbomam commented on .github/pull_request_template_berk.md:
```diff
@@ -0,0 +1,86 @@
+
+_If your change **does** results in downstream changes, provide a justification/need for this change_ >-
+
+_Examples of downstream changes include any change that requires an update or change to workflows, workflow automation, Postgres ingest, Jupyter notebooks, runtime, & more_
```
is a change to example data validation a downstream effect?

mslarae13 replied:
No I don't think so. But it's more about making sure people DO put in examples if they should.

eecavanna commented:
I also plan to propose some changes to some wording (I'll shoot for doing so within the next 24 hours).

turbomam commented on .github/ISSUE_TEMPLATE/pull_request_template_berk_schema.yml:
```diff
@@ -0,0 +1,103 @@
+        - label: Bug Fix
+        - label: Optimization
+        - label: Documentation Update
+        - label: Content change (Ex- slot or Class name, slot multiplicity changes, slot movement from one Class to another, creating a new slot/Class, Enum value changes)
```
by multiplicity do you mean `multivalued`? I suggest using real LinkML terminology.

mslarae13 commented on .github/ISSUE_TEMPLATE/pull_request_template_berk_schema.yml:
```diff
@@ -0,0 +1,103 @@
+    attributes:
+      label: Issues
+      description: What issues are part of this PR?
+      placeholder: ex. /issues/issues/100, nmdc-schema/issues/400
```
I used /issues/100 becuase the PR tempalte is in berk-schema but the issues are in nmdc-scheam/issues/ or issues/issues... they're not in berk & wanted you to be able to find the right repo/issue relationship

turbomam replied:
Could we use full URLs then?

mslarae13 replied:
Eric also supports doing full URL if it's across repos.
Will update

mslarae13 commented:
Note. .yml doesn't work for PRs! So switchin to markdown. Updated file link corrected above https://github.com/microbiomedata/berkeley-schema-fy24/blob/1995-add-pr-template_berk_schema/.github/pull_request_template_berk.md

mslarae13 commented:
From Mark via slack

Q: under what circumstances would somebody check off 'Schema change: Content or 'Schema change: Cleanup' but not 'Changes to any files or directories in src/schema or vice versa?
R1: let's use this terminology, capitalization and styling: name , multivalued, slot, class, enum, mappings . There's not such thing as a 'term' in LInkML or the nmdc-schema.
R2: instead of 'slot movement', let's ask about any changes to the ways slots are assigned to any classes. That would include additions and removals in addition to movement
R3: I recommend adding descriptions to these PR types:
Refactor
 Feature
 Optimization

mslarae13 commented:
> Q: under what circumstances would somebody check off 'Schema change: Content or 'Schema change: Cleanup' but not 'Changes to any files or directories in src/schema or vice versa?
- I don't think they should! But, I like having a check / confirmation. If preferred,, we can reformat to. But I think the point here is if there could be changes to src/schema that aren't one of those 2...and if you check the other 2, you should check src/schema. .... If we do make this format change.. we should add an "other" option

- [ ]  Changes to any files or directories in src/schema
  - [ ] Schema change: Content
    - slot or class name
    - slot multiplicity changes (from a single value to a list or vice versa)
    - slot movement from one class to another
    - creating a new slot/Class
    - Enum value changes
  - [ ] Schema change: Cleanup
    - descriptions of slots, classes, or enums
    - removal of unused, commented, or invalid code
    - updated mappings of terms, classes, or slots to ontologies
    - added an additional Enum to accommodate new or future metadata

> R1: let's use this terminology, capitalization and styling: name , multivalued, slot, class, enum, mappings . There's not such thing as a 'term' in LInkML or the nmdc-schema.
@turbomam can you put a comment in the PR line where you want this changed. I see term.. but i'm not sure what you mean

> R2: instead of 'slot movement', let's ask about any changes to the ways slots are assigned to any classes. That would include additions and removals in addition to movement
- Done, https://github.com/microbiomedata/berkeley-schema-fy24/pull/217/commits/62630db7d74de2b076a2089b9ae2a7770de74913

turbomam commented:
thanks @mslarae13 

>can you put a comment in the PR line where you want this changed

will do

> I don't think they should! But, I like having a check / confirmation

I like having lots of checks too, but not checks that could contradict each other. Any change to any file in `src/schema/` is either a cleanup action or a content change.

turbomam commented on .github/ISSUE_TEMPLATE/pull_request_template_berk_schema.yml:
```diff
@@ -0,0 +1,103 @@
+        # Guidelines
+        
+        ## _Soft_ Schema Freeze
+        The nmdc-schema and berkeley-schema are under a light freeze this means changes **should not** be made that have any down stream implications. 
```
soft

mslarae13 replied:
I thought I fixed that?!
thanks!

turbomam commented:
actually

>can you put a comment in the PR line where you want this changed

do you just want me to make the edits? they're scattered on many lines.

eecavanna commented on .github/PULL_REQUEST_TEMPLATE/pull_request_template_berk.md:
```diff
@@ -0,0 +1,86 @@
+- [ ] Bug Fix
+- [ ] Optimization
+- [ ] Documentation
+- [ ] Schema change: Content
+  - slot or class name
+  - slot multiplicity changes (from a single value to a list or vice versa)
+  - any changes to the ways slots are assigned to any classes
+  - creating a new slot/Class
+  - Enum value changes
+- [ ] Schema change: Cleanup
+  - descriptions of slots, classes, or enums
+  - removal of unused, commented, or invalid code
+  - updated mappings of terms, classes, or slots to ontologies
+  - added an additional Enum to accommodate new or future metadata
```
FYI: I'm working on a proposal for these two bullet points (and their contents). I'll post it here once I'm done.

eecavanna replied:
```suggestion
- [ ] Schema change — Structure and content; for example:
  - adds or renames a class, slot, or Enum
  - changes the multiplicity of a slot (e.g. string ↔ list)
  - changes how a slot is used by a class
  - adds or updates an Enum value
- [ ] Schema change — Cleanup and documentation; for example:
  - updates the description of a class, slot, or Enum
  - updates comments
  - deletes unused code
```

eecavanna replied:
I'll close this proposal since the code is was based on has been changed via a different means. I will re-express the proposal in a few minutes.

turbomam commented:
@eecavanna do you think it will be hard to coordinate with https://github.com/microbiomedata/berkeley-schema-fy24/pull/221

@mslarae13 I added this PR against your PR so you could see the differences

turbomam reviewed (commented):
I suggest auto-reformatting these Markdown source files. You can do that in PyCharm on Mac with command-option-L

turbomam reviewed (commented):
I think it would be good for us to look though these documents and make sure we are distinguishing between changes to `enum`s and changes to the `permissible_values` of an `enum`.

turbomam commented:
> Please identify the issue(s) below.

Ideally, the issue will appear in the right hand side of the PR screen. I think the section is called "development"

That link is created automatically when issues are linked to branches

turbomam commented on .github/PULL_REQUEST_TEMPLATE/pull_request_template_berk.md:
```diff
@@ -0,0 +1,86 @@
+
+## Related Issues
+
+> All PRs should relate to or fix an issue(s). Please identify the issue(s) below.
```
Ideally, the issue will appear in the right hand side of the PR screen. I think the section is called "development"

That link is created automatically when issues are linked to branches

eecavanna commented on .github/PULL_REQUEST_TEMPLATE/pull_request_template_berk.md:
```diff
@@ -0,0 +1,85 @@
+- [ ] Bug Fix
+- [ ] Optimization
+- [ ] Documentation
+- [ ] Schema change: Content
+  - `slot` or `class` name
+  - changes to a `slot`'s `multivalued` state. `true` means the schema expects a list of values and `false` means  a single value is expected.
+  - any changes to the ways `slot`s are assigned to any `class`es
+  - creating a new `slot` or `class`
+  - additions, removals or changes to the `permissible_values` in an `enum`
+- [ ] Schema change: Cleanup
+  - descriptions of `slot`s, `class`es, or `enum`s
+  - removal of unused, commented, or invalid code
+  - updated `mappings` of `slot`s, `class`es, or `slot`s to ontologies
+  - addition of a new `enum` in the absence of making that `enum` the `range` for any existing `slot`
```
I made two groups:

- Structure and content
- Cleanup and preparation

I also standardized the phrasing of the sub-bullets.

```suggestion
- [ ] Schema change: Structure and content
  - created, updated, or deleted a `class`, `slot`, or `enum`
  - changed whether a `slot` is `multivalued`
  - changed the way a `slot` is assigned to a `class`
  - changed the `permissible_values` of an `enum`
  - _etc._
- [ ] Schema change: Cleanup and preparation
  - updated the description of a `class`, `slot`, or `enum`
  - updated the `mappings` of a `class`, `slot`, or `enum` to an ontology
  - added an `enum` for _future_ use (it is not in the `range` of any `slot`)
  - _etc._
```

eecavanna replied:
Some of the attendees of today's metadata meeting discussed this suggestion today and nobody expressed objections to it.

Also, some of them clarified that they see the first group as "more likely to have downstream implications" and the second group "less likely to have downstream implications" (relative to the other group).

mslarae13 commented:
Can someone review so I can merge? :) @eecavanna  @turbomam

eecavanna commented:
I'm reviewing it now 🔍

eecavanna commented on .github/PULL_REQUEST_TEMPLATE/pull_request_template_berk.md:
```diff
@@ -0,0 +1,85 @@
+
+## Does this PR have any downstream implications?
+
+> **Ideally**, no schema changes will be performed that have downstream implications.
```
I propose removing this line as I think it's redundant with the intro. Also, phrasing of a "schema change" being "performed" (as opposed to "made" or "applied") doesn't sound natural to my ears.

eecavanna commented on .github/PULL_REQUEST_TEMPLATE/pull_request_template_berk.md:
```diff
@@ -0,0 +1,85 @@
+
+> **Ideally**, no schema changes will be performed that have downstream implications.
+>
+> Examples of downstream changes: any change that requires a change to workflows, workflow automation, the Mongo-to-Postgres ingest process, Jupyter notebooks, the Runtime, etc.
```
```suggestion
> Examples: any change that requires a change to workflows, workflow automation, the Mongo-to-Postgres ingest process, Jupyter notebooks, the Runtime, etc.
```

eecavanna commented on .github/PULL_REQUEST_TEMPLATE/pull_request_template_berk.md:
```diff
@@ -0,0 +1,85 @@
+>
+> Examples of downstream changes: any change that requires a change to workflows, workflow automation, the Mongo-to-Postgres ingest process, Jupyter notebooks, the Runtime, etc.
+
+- [ ] Yes _(Add a justification below)_
```
```suggestion
- [ ] Yes _(Explain below)_
```

eecavanna commented on .github/PULL_REQUEST_TEMPLATE/pull_request_template_berk.md:
```diff
@@ -0,0 +1,85 @@
+
+## Does this PR have any downstream implications?
+
+> **Ideally**, no schema changes will be performed that have downstream implications.
+>
```
```suggestion

```

I propose removing this as (1) I think it's redundant with the intro, and (2) saying a "schema change" will be "performed" (as opposed to "made" or "applied") doesn't sound natural to my ears.

eecavanna commented:
Hi @mslarae13, this looks good to me, other than the 3 things I commented on. If those get "accepted", I'll approve and merge (or you can merge at that point).

eecavanna commented:
And I haven't been reviewing the YAML file (I don't know what its role is, given that GitHub doesn't support using a YAML file as a PR template—whereas it does for an Issue template).

eecavanna reviewed (approved):
The Markdown file (i.e. PR template) looks good to me! Thanks for tackling this, @mslarae13.

---

Note: I did not review the YAML file as I don't know what its role is (given that GitHub doesn't support using a YAML file as a PR template).

eecavanna commented:
I deleted the YAML file as @mslarae13 and I discussed on Slack just now.

I'm comfortable with this PR branch being merged into the `main` branch.

mslarae13 commented:
merging. Any additional changes should be suggested via another branch & PR.

@eecavanna is going to add/update the branch & PR for nmdc-schema

eecavanna commented:
I finished updating the PR in `nmdc-schema` (i.e. this one: https://github.com/microbiomedata/nmdc-schema/pull/2105) to have the same PR template as the final one introduced by this PR branch.
